arch/sim/src/sim/up_i2cbuslinux.c

```diff
@@ -0,0 +1,310 @@
+  .reset = linux_i2cbus_reset,
+#endif
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: linux_i2cbus_reset
+ *
+ * Description:
+ *   Provide i2c reset
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_I2C_RESET
+static int linux_i2cbus_reset(struct i2c_master_s *dev)
+{
+  return -1; /* Not implemented */
+}
+#endif
+
+/****************************************************************************
+ * Name: linux_i2cbus_transfer
+ *
+ * Description:
+ *   Provide i2c transfer
+ *
+ ****************************************************************************/
+
+static int linux_i2cbus_transfer(struct i2c_master_s *dev,
+                                 struct i2c_msg_s *msgs, int count)
+{
+  int ret = 0;
+  struct linux_i2cbus_master_s *priv = (struct linux_i2cbus_master_s *)dev;
+  int file = priv->file;
+  uint8_t *pack_buf = NULL;
+  struct i2c_rdwr_ioctl_data ioctl_data;
+  struct i2c_msg l_msgs[2];  /* We only support up to 2 messages */
+  ioctl_data.msgs = l_msgs;
+
+  /* Many i2c bus do not play well with combined messages via the Linux
+   * interface, this makes stitching things together a little harder
+   * because NuttX provides the ability to hold the bus without ending
+   * with a STOP which is not ideal in general, and not possible with
+   * Linux.
+   */
+
+  if ((msgs[0].flags & NUTTX_I2C_M_TEN) || (msgs[1].flags & NUTTX_I2C_M_TEN))
+    {
+      /* Linux also has somewhat poor support for 10bit addresses and they
+       * are quite rare so we just don't support them for now here
+       */
+
+      return -1;
+    }
+
+  ioctl_data.msgs = l_msgs;
+
+  if (count == 1)
+    {
+      if (msgs[0].flags & NUTTX_I2C_M_NOSTOP || \
+          msgs[0].flags & NUTTX_I2C_M_NOSTART)
+        {
+          /* Do not support leaving the bus hanging or try to send
+           * without first starting.
+           */
+
+          return -1;
+        }
+
+      l_msgs[0].addr = msgs[0].addr;
+      l_msgs[0].buf = msgs[0].buffer;
+      l_msgs[0].len = msgs[0].length;
+      l_msgs[0].flags = 0;
+      if (msgs[0].flags & NUTTX_I2C_M_READ)
+        {
+           l_msgs[0].flags |= I2C_M_RD;
+        }
+
+      ioctl_data.nmsgs = 1;
+    }
+  else if(count == 2)
+    {
+      /* Addresses should be the same */
+
+      if (msgs[0].addr != msgs[1].addr)
+        {
+          return -1;
+        }
+
+      /* Check if we are about to do a read of a register
+       * NuttX interface represents this as WRITE(NOSTOP) + READ
+       * Linux interface represents this as READ + READ
```

Comment:
Is this comment really right? I never saw READ + READ mean to read of a register

Reply:
This is an abstraction, READ/WRITE in linux are a higher level so they get mapped down to what NuttX is doing.  This is correct and I have attached the waveforms that show this. 
See this kernel driver to better understand what is going on here:
https://elixir.bootlin.com/linux/latest/source/drivers/hid/hid-mcp2221.c#L273

Reply:
You can see how the READ+READ gets mapped to MCP2221_I2C_WR_NO_STOP - MCP2221_I2C_RD_RPT_START which is what NuttX has.

Reply:
Sorry my description is not quite right, but the implementation is correct here.

Reply:
Fixed the comment to reflect what is here:
NuttX:  WRITE(NOSTOP) + READ
Linux: READ + READ